File: src/problem15/Solution.java
public class Solution {
	public List<List<Integer>> threeSum(int[] nums) {
		List<List<Integer>> lists = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < len-2; i++) {
            if (nums[i] > 0 || nums[len-1] < 0)
                break;;
            if (i > 0 && nums[i] == nums[i-1])
                continue;
            int j = i + 1;
            int k = len - 1;
            while (j < k) {
                if (nums[k] < 0)
                    break;
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    List<Integer> list = new ArrayList<>();
                    list.add(nums[i]);
                    list.add(nums[j]);
                    list.add(nums[k]);
                    lists.add(list);
                    ++j;
                    --k;
                    while (j < k && nums[j] == nums[j-1])
                        ++j;
                    while (j < k && nums[k] == nums[k+1])
                        --k;
                }
                else if (sum < 0) {
                    ++j;
                }
                else {
                    --k;
                }
            }
        }
        return lists;
	}
}

File: src/problem15/Solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        if (nums.size() < 3)
            return result;
        vector<int> tmp(3);
        sort(nums.begin(), nums.end());
        auto last = nums.end();
        for (auto i = nums.begin(); i != last-2; i++) {
            if (i > nums.begin() && *i == *(i-1)) continue;
            auto j = i + 1;
            auto k = last - 1;
            while(j < k) {
                if (*i + *j + *k < 0) {
                    j++;
                    while(*j == *(j-1)) j++;
                } else if (*i + *j + *k > 0) {
                    k--;
                    while(*k == *(k+1)) k--;
                } else {
                    tmp[0] = *i;
                    tmp[1] = *j;
                    tmp[2] = *k;
                    result.push_back(tmp);
                    j++;
                    k--;
                    while(j<k && *j==*(j-1) && *k==*(k+1)) {
                        j++;
                        k--;
                    }
                }
            }
        }
    return result;
    }
};